akka-actor/src/main/scala/akka/pattern/Patterns.scala

```diff
@@ -454,6 +455,76 @@ object Patterns {
       value: CompletionStage[T]): CompletionStage[T] =
     afterCompletionStage(duration.asScala, scheduler)(value)(context)
 
+  /**
+   * Returns an internally retrying [[scala.concurrent.Future]]
+   * The first attempt will be made immediately, each subsequent attempt will be made immediately
+   * if the previous attempt failed.
+   * If attempts are exhausted the returned future is simply the result of invoking attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent tries and
+   * therefore must be thread safe (not touch unsafe mutable state).
+   */
+  def retry[T](attempt: Callable[Future[T]], attempts: Int, context: ExecutionContext): Future[T] = {
+    require(attempt != null, "Parameter attempt should not be null.")
+    scalaRetry(() => attempt.call, attempts)(context)
+  }
+
+  /**
+   * Returns an internally retrying [[java.util.concurrent.CompletionStage]]
+   * The first attempt will be made immediately, each subsequent attempt will be made immediately
+   * if the previous attempt failed.
+   * If attempts are exhausted the returned completion operator is simply the result of invoking attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent tries
+   * and therefore must be thread safe (not touch unsafe mutable state).
+   */
+  def retry[T](attempt: Callable[CompletionStage[T]], attempts: Int, ec: ExecutionContext): CompletionStage[T] = {
```

Comment:
These two methods will be erased the same and not possible to call from Java. Skip the `Future` one and keep the `CompletionStage` as it is the Java API?

Reply:
works fine

Reply:
It's not valid in to define statics like that in Java, it wouldn't compile at all actually,  there seem to be a loophole allowing it from Scala. Given that it would not be valid in Java we should not rely on that when adding new methods though. Especially when we should only have the CS signatures for Java anyway.

Reply:
I'm fine to only keep the CS one, and that's what I would like too, But the test show it can be called from Java.

Reply:
Yes, it currently can but as it is not possible to express the same thing (conflicting overload) in Java we'd like to avoid relying on that it keeps working in the future.